src/facade/dragonfly_connection.cc

```diff
@@ -823,64 +870,66 @@ void Connection::ConnectionFlow(FiberSocketBase* peer) {
+                                 absl::FunctionRef<MessageHandle()> dispatch_async) {
+  // Avoid sync dispatch if we can interleave with an ongoing async dispatch
+  bool can_dispatch_sync = !cc_->async_dispatch;
 
   // Avoid sync dispatch if we already have pending async messages or
-  // can potentially receive some (subscriptions > 0). Otherwise the dispatch
-  // fiber might be constantly blocked by sync_dispatch.
+  // can potentially receive some (subscriptions > 0)
   if (dispatch_q_.size() > 0 || cc_->subscriptions > 0)
     can_dispatch_sync = false;
 
-  if (can_dispatch_sync) {
-    ShrinkPipelinePool();  // Gradually release pipeline request pool.
-
-    RespExpr::VecToArgList(tmp_parse_args_, &tmp_cmd_vec_);
+  // Dispatch async if we're handling a pipeline or if we can't dispatch sync.
+  if (has_more || !can_dispatch_sync) {
+    SendAsync(dispatch_async());
 
+    if (dispatch_q_.size() > 10)
+      ThisFiber::Yield();
+  } else {
+    ShrinkPipelinePool();  // Gradually release pipeline request pool.
     {
       cc_->sync_dispatch = true;
-      service_->DispatchCommand(absl::MakeSpan(tmp_cmd_vec_), cc_.get());
+      dispatch_sync();
       cc_->sync_dispatch = false;
     }
-
     last_interaction_ = time(nullptr);
 
     // We might have blocked the dispatch queue from processing, wake it up.
     if (dispatch_q_.size() > 0)
       evc_.notify();
-
-  } else {
-    SendAsync(MessageHandle{FromArgs(std::move(tmp_parse_args_), heap)});
-    if (dispatch_q_.size() > 10)
-      ThisFiber::Yield();
   }
 }
 
 Connection::ParserStatus Connection::ParseRedis(SinkReplyBuilder* orig_builder) {
   uint32_t consumed = 0;
-
   RedisParser::Result result = RedisParser::OK;
-  mi_heap_t* tlh = mi_heap_get_backing();
+
+  // Re-use connection local resources to reduce allocations
+  RespVec& parse_args = tmp_parse_args_;
+  CmdArgVec& cmd_vec = tmp_cmd_vec_;
+
+  auto dispatch_sync = [this, &parse_args, &cmd_vec] {
+    RespExpr::VecToArgList(parse_args, &cmd_vec);
+    service_->DispatchCommand(absl::MakeSpan(cmd_vec), cc_.get());
+  };
+  auto dispatch_async = [this, &parse_args, tlh = mi_heap_get_backing()]() -> MessageHandle {
+    return {FromArgs(std::move(parse_args), tlh)};
+  };
 
   do {
-    result = redis_parser_->Parse(io_buf_.InputBuffer(), &consumed, &tmp_parse_args_);
+    result = redis_parser_->Parse(io_buf_.InputBuffer(), &consumed, &parse_args);
 
-    if (result == RedisParser::OK && !tmp_parse_args_.empty()) {
-      RespExpr& first = tmp_parse_args_.front();
-      if (first.type == RespExpr::STRING) {
+    if (result == RedisParser::OK && !parse_args.empty()) {
+      if (RespExpr& first = parse_args.front(); first.type == RespExpr::STRING)
         DVLOG(2) << "Got Args with first token " << ToSV(first.GetBuf());
-      }
 
-      DispatchCommand(consumed, tlh);
+      bool has_more = consumed < io_buf_.InputLen();
+
+      if (tl_traffic_logger.log_file)  // Log command as soon as we receive it
```

Comment:
We didn't log anything for memcached before. It's also not a "one liner" because we have to serialize our parsed command back, including all the flags